File: guardian-agent-apk/src/com/fleetops/guardian/EnrollmentActivity.java
package com.fleetops.guardian;

import android.app.Activity;
import android.app.admin.DevicePolicyManager;
import android.content.*;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;
import android.widget.*;
import org.json.JSONObject;
import com.fleetops.guardian.R;

public class EnrollmentActivity extends Activity {
    private static final int REQ_PERMISSIONS = 101;
    private static final int REQ_ADMIN       = 102;

    private EditText   etServerUrl, etOrgToken, etDeviceName;
    private Button     btnEnroll, btnPaste;
    private ProgressBar progressBar;
    private TextView   tvStatus;
    private DevicePrefs prefs;

    @Override
    protected void onCreate(Bundle saved) {
        super.onCreate(saved);
        prefs = new DevicePrefs(this);

        if (prefs.isEnrolled()) {
            requestPermissionsAndAdmin();
            return;
        }

        setContentView(R.layout.activity_enrollment);
        etServerUrl  = (EditText)    findViewById(R.id.etServerUrl);
        etOrgToken   = (EditText)    findViewById(R.id.etOrgToken);
        etDeviceName = (EditText)    findViewById(R.id.etDeviceName);
        btnEnroll    = (Button)      findViewById(R.id.btnEnroll);
        btnPaste     = (Button)      findViewById(R.id.btnPaste);
        progressBar  = (ProgressBar) findViewById(R.id.progressBar);
        tvStatus     = (TextView)    findViewById(R.id.tvStatus);

        btnEnroll.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { doEnroll(); }
        });
        btnPaste.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { pasteFromClipboard(); }
        });
    }

    // ── Clipboard quick-enroll ────────────────────────────────────────────────

    private void pasteFromClipboard() {
        ClipboardManager cm = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
        if (cm == null || !cm.hasPrimaryClip()) {
            status("Clipboard is empty", 0xFFFF3355);
            return;
        }
        ClipData.Item item = cm.getPrimaryClip().getItemAt(0);
        if (item == null) return;
        String text = item.getText() != null ? item.getText().toString().trim() : "";
        if (text.isEmpty()) { status("Clipboard is empty", 0xFFFF3355); return; }
        try {
            JSONObject json = new JSONObject(text);
            String url   = json.optString("url", "");
            String token = json.optString("token", "");
            String name  = json.optString("name", "");
            if (!url.isEmpty())   etServerUrl.setText(url);
            if (!token.isEmpty()) etOrgToken.setText(token);
            if (!name.isEmpty())  etDeviceName.setText(name);
            status("Config pasted — tap Enroll to continue", 0xFF00FF88);
        } catch (Exception e) {
            status("Clipboard doesn't contain a valid config JSON", 0xFFFF3355);
        }
    }

    // ── Enrollment ────────────────────────────────────────────────────────────

    private void doEnroll() {
        String url   = etServerUrl.getText().toString().trim();
        String token = etOrgToken.getText().toString().trim();
        String name  = etDeviceName.getText().toString().trim();

        if (url.isEmpty())   { status("Enter server URL", 0xFFFF3355); return; }
        if (token.isEmpty()) { status("Enter org token",  0xFFFF3355); return; }

        while (url.endsWith("/")) url = url.substring(0, url.length() - 1);
        if (name.isEmpty()) name = android.os.Build.MANUFACTURER + " " + android.os.Build.MODEL;
        if (name.trim().isEmpty()) name = "Guardian Device";

        final String finalUrl   = url;
        final String finalToken = token;
        final String finalName  = name;
        final String imei       = getDeviceId();

        setLoading(true, "Enrolling...");

        new Thread(new Runnable() {
            @Override
            public void run() {
                final ApiClient.EnrollResult result =
                    ApiClient.enroll(finalUrl, finalToken, finalName, imei);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        setLoading(false, null);
                        if (result.token != null) {
                            prefs.saveEnrollment(finalUrl, result.token, finalName);
                            status("Enrolled — setting up device protection...", 0xFF00FF88);
                            requestPermissionsAndAdmin();
                        } else {
                            status(result.error != null ? result.error : "Enrollment failed",
                                0xFFFF3355);
                        }
                    }
                });
            }
        }).start();
    }

    // ── Permission + Admin flow ───────────────────────────────────────────────

    private void requestPermissionsAndAdmin() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            java.util.List<String> needed = new java.util.ArrayList<>();
            if (checkSelfPermission(android.Manifest.permission.ACCESS_FINE_LOCATION)
                    != PackageManager.PERMISSION_GRANTED) {
                needed.add(android.Manifest.permission.ACCESS_FINE_LOCATION);
            }
            if (checkSelfPermission(android.Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                needed.add(android.Manifest.permission.CAMERA);
            }
            if (!needed.isEmpty()) {
                requestPermissions(needed.toArray(new String[0]), REQ_PERMISSIONS);
                return;
            }
        }
        requestDeviceAdmin();
    }

    @Override
    public void onRequestPermissionsResult(int code, String[] perms, int[] grants) {
        // Continue regardless of result — request device admin next
        requestDeviceAdmin();
    }

    private void requestDeviceAdmin() {
        DevicePolicyManager dpm =
            (DevicePolicyManager) getSystemService(DEVICE_POLICY_SERVICE);
        ComponentName admin = new ComponentName(this, GuardianAdminReceiver.class);
        if (dpm != null && !dpm.isAdminActive(admin)) {
            Intent intent = new Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN);
            intent.putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN, admin);
            intent.putExtra(DevicePolicyManager.EXTRA_ADD_EXPLANATION,
                "FleetOps Guardian requires device administrator access to protect " +
                "this device and prevent unauthorized removal by field personnel.");
            startActivityForResult(intent, REQ_ADMIN);
        } else {
            launch();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQ_ADMIN) {
            // Launch regardless — admin activation is strongly recommended but not blocking
            launch();
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private String getDeviceId() {
        try {
            String id = android.provider.Settings.Secure.getString(
                getContentResolver(), android.provider.Settings.Secure.ANDROID_ID);
            return id != null ? id : "";
        } catch (Exception e) {
            return "";
        }
    }

    private void setLoading(boolean loading, String msg) {
        btnEnroll.setEnabled(!loading);
        progressBar.setVisibility(loading ? View.VISIBLE : View.GONE);
        if (msg != null) status(msg, 0xFF4A7090);
    }

    private void status(String msg, int color) {
        tvStatus.setText(msg);
        tvStatus.setTextColor(color);
    }

    private void launch() {
        startActivity(new Intent(this, MainActivity.class));
        finish();
    }
}

File: guardian-agent/app/src/main/java/com/fleetops/guardian/ui/enrollment/EnrollmentActivity.kt
package com.fleetops.guardian.ui.enrollment

import android.Manifest
import android.app.admin.DevicePolicyManager
import android.content.ComponentName
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.fleetops.guardian.BuildConfig
import com.fleetops.guardian.data.prefs.DevicePrefs
import com.fleetops.guardian.databinding.ActivityEnrollmentBinding
import com.fleetops.guardian.receiver.GuardianDeviceAdminReceiver
import com.fleetops.guardian.service.GuardianService
import com.fleetops.guardian.ui.main.MainActivity
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class EnrollmentActivity : AppCompatActivity() {

    @Inject lateinit var devicePrefs: DevicePrefs

    private lateinit var binding: ActivityEnrollmentBinding
    private val viewModel: EnrollmentViewModel by viewModels()

    private val locationPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        val fineGranted = permissions[Manifest.permission.ACCESS_FINE_LOCATION] == true
        val coarseGranted = permissions[Manifest.permission.ACCESS_COARSE_LOCATION] == true
        if (!fineGranted && !coarseGranted) {
            Toast.makeText(this, "Location permission is required for fleet tracking", Toast.LENGTH_LONG).show()
        }
    }

    private val notificationPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { /* handled silently */ }

    private val deviceAdminLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { navigateToMain() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityEnrollmentBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Hide form until enrollment check completes — prevents flash on already-enrolled devices
        binding.enrollForm.visibility = View.INVISIBLE

        lifecycleScope.launch {
            if (devicePrefs.isEnrolled()) {
                navigateToMain()
            } else {
                binding.enrollForm.visibility = View.VISIBLE
                setupUi()
                observeViewModel()
                requestRequiredPermissions()
            }
        }
    }

    private fun setupUi() {
        binding.etServerUrl.setText(BuildConfig.SERVER_URL)

        binding.btnEnroll.setOnClickListener {
            viewModel.enroll(
                serverUrl = binding.etServerUrl.text.toString(),
                orgToken = binding.etOrgToken.text.toString(),
                deviceName = binding.etDeviceName.text.toString(),
            )
        }
    }

    private fun observeViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state -> handleUiState(state) }
            }
        }
    }

    private fun handleUiState(state: EnrollmentUiState) {
        when (state) {
            is EnrollmentUiState.Idle -> {
                setLoading(false)
                binding.tvStatus.visibility = View.GONE
            }
            is EnrollmentUiState.Loading -> {
                setLoading(true)
                binding.tvStatus.visibility = View.VISIBLE
                binding.tvStatus.text = "Enrolling device..."
                binding.tvStatus.setTextColor(ContextCompat.getColor(this, android.R.color.white))
            }
            is EnrollmentUiState.Success -> {
                setLoading(false)
                binding.tvStatus.visibility = View.VISIBLE
                binding.tvStatus.text = "Enrollment successful! Activating security..."
                binding.tvStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_light))
                requestDeviceAdmin()
            }
            is EnrollmentUiState.Error -> {
                setLoading(false)
                binding.tvStatus.visibility = View.VISIBLE
                binding.tvStatus.text = "Error: ${state.message}"
                binding.tvStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_light))
            }
        }
    }

    private fun setLoading(loading: Boolean) {
        binding.progressBar.visibility = if (loading) View.VISIBLE else View.GONE
        binding.btnEnroll.isEnabled = !loading
        binding.etServerUrl.isEnabled = !loading
        binding.etOrgToken.isEnabled = !loading
        binding.etDeviceName.isEnabled = !loading
    }

    private fun navigateToMain() {
        GuardianService.startService(this)
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

    private fun requestDeviceAdmin() {
        val dpm = getSystemService(DEVICE_POLICY_SERVICE) as DevicePolicyManager
        val component = ComponentName(this, GuardianDeviceAdminReceiver::class.java)
        if (dpm.isAdminActive(component)) {
            navigateToMain()
            return
        }
        val intent = Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN).apply {
            putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN, component)
            putExtra(DevicePolicyManager.EXTRA_ADD_EXPLANATION,
                "Required to protect fleet data, prevent app removal, and enable remote device management.")
        }
        deviceAdminLauncher.launch(intent)
    }

    private fun requestRequiredPermissions() {
        val permissionsToRequest = mutableListOf<String>()
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            permissionsToRequest.add(Manifest.permission.ACCESS_FINE_LOCATION)
            permissionsToRequest.add(Manifest.permission.ACCESS_COARSE_LOCATION)
        }
        if (permissionsToRequest.isNotEmpty()) {
            locationPermissionLauncher.launch(permissionsToRequest.toTypedArray())
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                notificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
        }
    }
}
